### ISSUE TITLE
Explore provisioner compatibility with OpenTofu

### ISSUE THREAD

**Issue by bpmct** [coder/coder#12009]
We're an official [Hahsicorp Partner](https://www.hashicorp.com/partners/tech/coder#all) and are unimpacted by Hashicorp's [BSL license change](https://www.hashicorp.com/blog/hashicorp-adopts-business-source-license). However, our product aims to be neutral to clouds, deployment methods, and even provisioners (while we only support Terraform at this time, we intentionally build an abstract [provisioner](https://github.com/coder/coder/tree/main/provisioner) layer).

It could be beneficial to understand and document whether we support OpenTofu binaries if folks would prefer to use that over the official Hashicorp binary.

**Comment by coadler** [comment]
> I believe the hardcoded version check is somewhat arbitrary, but ensures compatibility with our example templates.

The version constraint isn't purely to prevent us from using the BUSL Terraform versions, but mainly to prevent using newer Terraform versions than we've tested ourselves. Chatted with @ammario and @bpmct the other week and decided to finally move to the BUSL versions of Terraform, so we should be moving off of the EOL 1.5.x release soon.

**Comment by bpmct** [comment]
> bumping the max terraform version in Coder

This is the part that will probably be difficult for users as it involves modifying Coder's source code (albeit in a simple way), creating a build, and maintaining a "fork/patch" of Coder that just bumps that max version. 

> so we should be moving off of the EOL 1.5.x release soon

Nice, so we can write a short guide on using Coder+OpenTofu as soon as we do this and bump our [maxTerraformVersion](https://github.com/coder/coder/blob/f57ce97b5aadd825ddb9a9a129bb823a3725252b/provisioner/terraform/install.go#L22-L25)

**Comment by mtojek** [comment]
> Nice, so we can this can be easier for users as soon as we do this and bump our [maxTerraformVersion](https://github.com/coder/coder/blob/f57ce97b5aadd825ddb9a9a129bb823a3725252b/provisioner/terraform/install.go#L22-L25)

Isn't there any problems around linking it with homebrew?

**Comment by johnstcn** [comment]
One other option we have here is to make the max terraform version constraint a soft constraint instead of a hard one.

Currently the logic is:
```
if terraform binary is present and terraform version is between minTerraformVersion and maxTerraformVersion:
  use existing terraform binary
else:
  install default terraform version
```

We could potentially change this to:
```
if terraform binary is present and terraform version is between minTerraformVersion and maxTerraformVersion:
  use existing terraform binary
else if terraform versions is greater than maxTerraform version:
  use existing terraform binary but warn about potential incompatibilities
else:
  install default terraform version
```

WDYT?

**Comment by bpmct** [comment]
> else if terraform versions is greater than maxTerraform version: use existing terraform binary but warn about potential incompatibilities

I like this. Can we also apply it to the minimum version as well? Perhaps some folks will want to stay on the non-BSL version of the official Terraform even after we stop "supporting" it

**Comment by mtojek** [comment]
Don't forget that this is a fork, so at some point APIs of both products may diverge, and we will need to treat `opentofu` as another provider.

**Comment by johnstcn** [comment]
> Don't forget that this is a fork, so at some point APIs of both products may diverge, and we will need to treat `opentofu` as another provider.

Correct. Right now the only way to use OpenTofu is to symlink `tofu` -> `terraform` so there's no chance of this happening by accident.

**Comment by kvendingoldo** [comment]
At this moment if you need to have a support of Terraform as well as OpenTofu (and Terragrunt :) ) in one tool you can use https://github.com/tofuutils/tenv which my team wrote some months ago. A lot of users switched to that tool to unify version management in the world of Terraform. 

You're welcome to open any issues or contribute to [tenv](https://github.com/tofuutils/tenv).

**Comment by michaelbrewer** [comment]
Also as Coder is FOSS product at its core, to support OpenTofu allow people to continue to use Coder without worrying about BSL. Of coz getting Pulumi is a bonus for sure.

**Comment by strike** [comment]
When migrating from Terraform to Tofu, keep in mind that the pinning in `.terraform.lock.hcl` will be affected.

For example, the `.terraform.lock.hcl` for Terraform:
```hcl
provider "registry.terraform.io/coder/coder" {
}
```

For Tofu:
```hcl
provider "registry.opentofu.org/coder/coder" {
}
```

**Comment by zkdzegede** [comment]
Has anything changed? Can we get an update on the steps to use OpenTofu in 2025.
